calm_adapter/calm_deletion_checker/src/main/scala/uk/ac/wellcome/platform/calm_deletion_checker/DeletionChecker.scala

```diff
@@ -1,14 +1,81 @@
 package uk.ac.wellcome.platform.calm_deletion_checker
 
-import uk.ac.wellcome.platform.calm_api_client.CalmApiClient
+import uk.ac.wellcome.platform.calm_api_client.{
+  CalmApiClient,
+  CalmQuery,
+  CalmSession
+}
 import weco.catalogue.source_model.CalmSourcePayload
 
-import scala.concurrent.Future
+import scala.concurrent.{ExecutionContext, Future}
+
+trait DeletionChecker {
+  type Record = CalmSourcePayload
+  type Records = Set[Record]
+
+  protected def nDeleted(records: Records): Future[Int]
+
+  // The implementation here is based on the algorithm given in Wang et al (2017)
+  // Found at: https://arxiv.org/abs/1407.2283, see `Algorithm 1`
+  //
+  // While there are more optimal algorithms that exist, the implementation of this one is very simple
+  def deletedRecords(batch: Records)(
+    implicit ec: ExecutionContext): Future[Records] = {
+    def findDeleted(records: Records, d: Int): Future[Records] = d match {
+      case 0                      => Future.successful(Set.empty)
+      case n if n == records.size => Future.successful(records)
+      case _ =>
+        val testSet = records.take(M(records.size, d))
+        for {
+          d1 <- nDeleted(testSet)
```

Comment:
So this works because something like: deletions are rare, so pick bigger sets of records to dismiss + do something clever to work out the optimal size based on how many deletions you see.

Reply:
Yep, that's a good summary. This algorithm being "nested" is a constraint that the authors (or rather, a previous paper) made up to make the problem more tractable, it's something along the lines of only recursing through strict subsets.

[This paper](http://www.learningtheory.org/colt2009/papers/004.pdf) is truly optimal and is very clever indeed, I would have liked to use it and did actually implement a good chunk of it but when I got to writing a fast Fourier transform in Scala I realised what I was doing and went back to the simpler one.